### DISCUSSION TITLE
How to add referrer in an ionic capacitor app

### DISCUSSION THREAD

**Discussion by manish1408** [Web3Auth/Web3Auth#396] [Help]
I am using web3auth in an ionic capacitor app. However, the redirect popup does not work and shows up blank.
On checking the console logs with the iOS simulator it shows OpenLogin - RPC Error: redirectUrl is not on origin. 

I have tried to whitelist the below URLs:
<bundle>://auth
ionic://auth
https://app.openlogin.com
ionic:/
capacitor://localhost
capacitor://auth

But none of these work
![Screenshot 2022-08-10 at 3 32 31 PM](https://user-images.githubusercontent.com/17041207/183874584-ff59ce9d-878c-498c-99c6-78ed71aa75b1.png)

How should we add a referer or add a redirectURL?

**Comment by dhruv500** [comment]
Hey @manish1408 your whitelist url will be your bundle Identifier for your iOS app followed by ://auth so something like this (xyz://auth) where xyz is your bundle Identifier which can be found in Targets -> General

**Reply by manish1408** [reply]
Hi @dhruv500  
We have already tried with our bundle identifier. it is ---->  com.f***c**ve://auth

But it doesn't work

**Reply by dhruv500** [reply]
Is the auth page opening inside of your app or in an external browser ?

**Reply by manish1408** [reply]
External browser.

**Reply by dhruvTor** [reply]
Please try adding your callback scheme in your url schemes in your ios app

**Reply by dhruvTor** [reply]
com.f*cve

**Reply by dziedziela** [reply]
tried that too and didn't work

**Reply by dziedziela** [reply]
![Screenshot 2022-08-12 at 12 14 26](https://user-images.githubusercontent.com/2911850/184343171-3e1aee22-cdd7-4b23-8aaf-92521e0af21e.png)

**Reply by dziedziela** [reply]
is there a way to pass the origin URL manually from code?

**Reply by dhruvTor** [reply]
Could you give me access to the code so that i can check exactly what's the issue

**Reply by dziedziela** [reply]
Would it be possible to schedule a call so we can screen share the existing solution as there are quite a few parts to consider?

**Reply by dhruvTor** [reply]
sure
https://calendly.com/dhruv-jaiswal

**Reply by dziedziela** [reply]
Booked for Monday, thank you!

**Comment by himanshuchawla009** [comment]
Openlogin cannot verify mobile origins, so you need to host a webpage for your app which and pass the URL of that webpage in the openloginAdapter loginSettings param as shown below and from you webpage, you can redirect to your mobile app.

The flow will look like this:-

Your app (using web3auth sdk) =>(Redirects to) => Openlogin.com (user will complete login here) => (After login, Redirects to your webpage) => Your hosted webpage =>(Redirects to) => Your IONIC  app

Sample Code:-

```

const openloginAdapter = new OpenloginAdapter({
"loginSettings": {
   "redirectUrl": "your web page url"
}
})

web3auth.configureAdapter(openloginAdapter)

```


Note: You will need to whitelist your webpage url rather thn mobile app uri in this case.

**Reply by manish1408** [reply]
Hello, @himanshuchawla009 Thank you for helping us out.

I am facing yet another issue after the redirect. The popup on the page after the redirect does not seem to close automatically. I have to exit the app and click on the login button again to get the wallet connected.

Do you have any idea about this behavior?

I am getting this response after the redirect https://domain.com/assets/redirect.html#result=eyJzdG9yZSI6eyJ0b3VjaElEUHJlZmVyZW5jZSI6ImRpc2FibGVkIiwiYXBwU3RhdGUiOiIiLCJlbWFpbCI6Im1nYXV0YW1AZmFuY3VydmUuY28iLCJhZ2dyZWdhdGVWZXJpZmllciI6InRrZXktZ29vZ2xlIiwibmFtZSI6Ik1hbmlzaCBHYXV0YW0iLCJwcm9maWxlSW1hZ2UiOiJodHRwczovL2xoMy5nb29nbGV1c2VyY29udGVudC5jb20vYS0vQUZkWnVjcklqSXBrM0tEbThQSW0wUjE5bDMtb25vS0pKZ2kwV1h3czhfbkg9czk2LWMiLCJ0eXBlT2ZMb2dpbiI6Imdvb2dsZSIsInZlcmlmaWVyIjoidG9ydXMiLCJ2ZXJpZmllcklkIjoibWdhdXRhbUBmYW5jdXJ2ZS5jbyIsImRhcHBTaGFyZSI6IiIsIm9BdXRoSWRUb2tlbiI6IiIsIm9BdXRoQWNjZXNzVG9rZW4iOiIiLCJpZFRva2VuIjoiZXlKaGJHY2lPaUpGVXpJMU5pSXNJblI1Y0NJNklrcFhWQ0lzSW10cFpDSTZJbFJaVDJkblh5MDFSVTlGWW14aFdTMVdWbEpaY1ZaaFJFRm5jSFJ1Wmt0V05EVXpOVTFhVUVNd2R6QWlmUS5leUpwWVhRaU9qRTJOakUwTVRRME1UTXNJbUYxWkNJNklrSkdOelZpVW5sdmRIQmpRMVkxZUVOUFEzQnNUR1JxV0RWV05WVkpTbmRWTWpORFZYQjVWWGxsU1daVE1tSmlTRVpwUWtGbFpuUjBUV1o0ZUdaMlkycEZSblIwZEdSMVEwcEtObTFwWnpObFZsaHZkWEo2TkNJc0ltNXZibU5sSWpvaU1ESmtZMll4TlRCaVpUbGlOelF3TkdJNU9EZGtOalE1WlRRNU1ESXhORFpsWWpWaU9EUmtPVEJqWkRFMU5ETTBORGRrTmpoaU1ETTVaalF3TUdJMFpERXlJaXdpYVhOeklqb2lhSFIwY0hNNkx5OWhjR2t1YjNCbGJteHZaMmx1TG1OdmJTSXNJbmRoYkd4bGRITWlPbHQ3SW5CMVlteHBZMTlyWlhraU9pSXdNalJrWTJNMU1XSTFOemcwWVRKaU1EYzBZalpsTVRGaE0yTTJOMlJoTUdKa016WXlPR1l3TTJZd1pXUmlNelF5WVRZM1ltVmpZVEZtT0RneU56TTBPVGNpTENKMGVYQmxJam9pZDJWaU0yRjFkR2hmWVhCd1gydGxlU0lzSW1OMWNuWmxJam9pYzJWamNESTFObXN4SW4xZExDSmxiV0ZwYkNJNkltMW5ZWFYwWVcxQVptRnVZM1Z5ZG1VdVkyOGlMQ0p1WVcxbElqb2lUV0Z1YVhOb0lFZGhkWFJoYlNJc0luQnliMlpwYkdWSmJXRm5aU0k2SW1oMGRIQnpPaTh2YkdnekxtZHZiMmRzWlhWelpYSmpiMjUwWlc1MExtTnZiUzloTFM5QlJtUmFkV055U1dwSmNHc3pTMFJ0T0ZCSmJUQlNNVGxzTXkxdmJtOUxTa3BuYVRCWFdIZHpPRjl1U0Qxek9UWXRZeUlzSW5abGNtbG1hV1Z5SWpvaWRHOXlkWE1pTENKMlpYSnBabWxsY2tsa0lqb2liV2RoZFhSaGJVQm1ZVzVqZFhKMlpTNWpieUlzSW1GblozSmxaMkYwWlZabGNtbG1hV1Z5SWpvaWRHdGxlUzFuYjI5bmJHVWlMQ0psZUhBaU9qRTJOakUxTURBNE1UTjkucTJwd2hGbVN5UnF0dlpfTHF1Q2VndTZjdkN5SUZpQ21Jc1p1Ymx3Y09xMldWNXpQcWxKTkxEZWtHVFJGeHN6UDc1YjZ6LVhqQmhCSHI1ZXM5c1JyOHcifSwicHJpdktleSI6IjI0MjEzYWQyNWI2NDdiNDIwOGRhMjFlNWE4OTFlNjMwYmRlNjgwMDlkMTYyODFlNDc2ODlhYThmMDFlMjI1NzciLCJlZDI1NTE5UHJpdktleSI6IjI0MjEzYWQyNWI2NDdiNDIwOGRhMjFlNWE4OTFlNjMwYmRlNjgwMDlkMTYyODFlNDc2ODlhYThmMDFlMjI1NzdmMzcyODljOWNlMGJlZjA0N2VhMGViZWI0MzEwZTA4Y2Y0ZTkxOWE4NDhkYzYwOTA2YTE0NGZlNjYyMTg3YTAxIn0&_pid=5447a75f367a8b8cfb8e874749d14a204372e69270ed17a999b563235ae87046&sessionId=928b23e66e7fda95d1bc66116d94b6724f6c0179f46078904d9e23cf389ef1aa

**Reply by BennyG93** [reply]
> The flow will look like this:-
> 
> Your app (using web3auth sdk) =>(Redirects to) => Openlogin.com (user will complete login here) => (After login, Redirects to your webpage) => Your hosted webpage =>(Redirects to) => Your IONIC app

@himanshuchawla009 Do you have an example of what this redirection code might look like? 

I am having a similar problem, we created a web app using Nextjs and have used this as a base for a capacitor mobile application. We are using Web3Auth Modal web SDK in `redirection` mode. The authentication flow is working fine, however when the user is redirected back to the application URL, it opens in the mobile browser instead of reopening the app.

Our code:
```
const openloginAdapter = new OpenloginAdapter({
      adapterSettings: {
        network: network,
        clientId: WEB3AUTH_CLIENT_ID,
        uxMode: 'redirect',
        redirectUrl: "https://example.com/dashboard",
     },
});
```

Capacitor has some documentation about deep-linking between web and mobile apps which seems similar to the flow you describe -> https://capacitorjs.com/docs/guides/deep-links
